modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -326,29 +298,83 @@ ThreadPool getThreadPool() {
+        }
 
-        final AtomicLong postCount = new AtomicLong();
+        RequestMetricCollector getMetricCollector(Operation operation, Purpose purpose) {
+            final String key;
+            if (purpose == Purpose.GENERIC) {
+                key = operation.value;
+            } else {
+                key = operation.value + "/" + purpose.getValue();
+                if (false == collectors.containsKey(key)) {
+                    collectors.put(key, buildMetricCollector(operation));
+                }
+            }
 
-        final AtomicLong deleteCount = new AtomicLong();
+            return collectors.get(key);
+        }
 
-        final AtomicLong abortCount = new AtomicLong();
+        Map<String, Long> statsMap() {
+            return collectors.entrySet()
+                .stream()
+                .collect(Collectors.toUnmodifiableMap(Map.Entry::getKey, entry -> entry.getValue().counter.get()));
+        }
 
-        Map<String, Long> toMap() {
-            final Map<String, Long> results = new HashMap<>();
-            results.put("GetObject", getCount.get());
-            results.put("ListObjects", listCount.get());
-            results.put("PutObject", putCount.get());
-            results.put("PutMultipartObject", postCount.get());
-            results.put("DeleteObjects", deleteCount.get());
-            results.put("AbortMultipartObject", abortCount.get());
-            return results;
+        IgnoreNoResponseMetricsCollector buildMetricCollector(Operation operation) {
+            return new IgnoreNoResponseMetricsCollector() {
+                @Override
+                public void collectMetrics(Request<?> request) {
+                    if (Assertions.ENABLED) {
+                        // assertions based on operation
+                        switch (operation) {
+                            case GET_OBJECT, LIST_OBJECTS -> {
+                                assert request.getHttpMethod().name().equals("GET");
+                            }
+                            case PUT_OBJECT -> {
+                                assert request.getHttpMethod().name().equals("PUT");
+                            }
+                            case PUT_MULTIPART_OBJECT -> {
+                                assert request.getHttpMethod().name().equals("PUT") || request.getHttpMethod().name().equals("POST");
+                            }
+                            case DELETE_OBJECTS -> {
+                                assert request.getHttpMethod().name().equals("POST");
+                            }
+                            case ABORT_MULTIPART_OBJECT -> {
+                                assert request.getHttpMethod().name().equals("DELETE");
+                            }
+                            default -> throw new AssertionError("unknown operation category " + operation);
+                        }
+                    }
```

Comment:
Nit: I'd rather things like this were a separate `boolean` method, called as `assert assertConsistentMethod(operation, request);`, instead of an explicit check for `Assertions.ENABLED`.

Reply:
Yep updated as suggested 8eada5ddb82fb8d78c682fcf0ee9dc14ac3f323e